### ISSUE TITLE
Aggregated report for multi-module projects

### ISSUE THREAD

**Issue by hcoles** [hcoles/pitest#41]
_From [msz...@wp.pl](https://code.google.com/u/116891989919461415994/) on April 26, 2012 14:34:14_

I would be useful to have a better support for multi-module projects. Currently it is possible to run pitest on multi-module project (tested with Maven plugin), but there are missing for me aggregated statistics.

Because it could be time consuming to implement at least links from a parent report to the child reports would be useful (now it is required to find reports for every module separately).

Probably others will have some other ideas about multi-module builds.

_Original issue: http://code.google.com/p/pitestrunner/issues/detail?id=41_

**Comment by szpak** [comment]
New link to Sonar plugin for PIT - https://github.com/SonarCommunity/sonar-pitest (as Codehaus reached EOL),

**Comment by michelmeeuwissen** [comment]
I created a small project to concat all generated .html reports to one html and publish it to a specific location (for example your webserver):

There is also a shell script included which can be used by Jenkins.

https://github.com/michelmeeuwissen/PITReportCombiner


_Remember that it this tool is made for a very specific requirement._

**Comment by jakub-bochenski** [comment]
Any word if/when this is going to implemented?

Also I'd claim this is a significant defect (not a "nice to have") viz how maven plugins are supposed to work with aggregate modules.

**Comment by galovics** [comment]
Any update here?

In my opinion, many projects have multi-module setup so I agree that this is not a "nice to have" feature but a must.

**Comment by KyleRogers** [comment]
@jakub-bochenski @galovics 
I can't agree that this is a defect. I think that's a useful enhancement but not an error. I'm  using the Sonar plugin mentioned by @szpak successfully. Thus I had no need for an aggregate mojo yet.

This is an open source project managed by @hcoles in his free time and supported by a small community of volunteers. If this is a "significant defect" for you pull requests are really welcome. Feel free to give it a try and let's see if we can support you if necessary.

**Comment by galovics** [comment]
The sonar plugin is an option. I also tried it out but didn't fit into the setup.

Don't get offended. Noone is here to blame someone that this is a missing feature. I just wanted to refresh the thread. :-)

**Comment by vmassol** [comment]
> Currently it is possible to run pitest on multi-module project (tested with Maven plugin)

Actually it's not working for me, I've raised https://github.com/hcoles/pitest/issues/314

**Comment by monperrus** [comment]
+1, an aggregated report would be so useful

**Comment by oscarlvp** [comment]
I have been working in a way to overcome this issue. What I'm doing is to use Maven to gather all dependencies from all sub-projects and conform the corresponding classpaths, that is mutable code, test code and all external dependencies. It works but has two main drawbacks. First it ignores surefire configurations that could be useful to detect disabled tests, but I think this can be also detected with Maven. The second drawback is that, if the test needs to be aware of the project folder structure, to access some input files, for example, it will fail, as all project structure is lost. The workaround right now is to replicate the project structure in another location and make this new location the working directory. It seems that in order to support this by code will require a huge amount of effort to make each test class aware of their working directory.

**Comment by hcoles** [comment]
@oscarlvp I think the correct solution is to do this in two stages.

1. Store data to disk for each project in xml/json/whatever (already happens, but see notes below)
2. Load and aggregate the data to a single html (or xml or json) file

The thing blocking this is that pitest does not currently write the coverage data out in a useful fashion. This html report needs this in addition to the mutation data.

Pitest does have support for writing coverage data to disk, but this is mainly for debug purposes. 

If the coverage data was available it shouldn't be to hard to create the aggregate report - the existing listeners could be used, but the data would be sourced from disk rather than the on the fly analysis.

But first the coverage data needs to be available as either

1. Additional data in the existing xml report file or
2. A separate well thought out file

**Comment by oscarlvp** [comment]
@hcoles If I understood well, this solution will target the issue of creating the report but doesn't target the fact that tests in a subproject may cover code in a different subproject, which happens very often and there are even projects that would put all tests in a separated subproject. The workaround I have used is able to create the aggregate report for the full project without changing the way reporting is done right now.

**Comment by sywor** [comment]
Adding myself to the list of people that wold like to see this feature.

**Comment by sarahBuisson** [comment]
+1

**Comment by stephen16cse** [comment]
+1

**Comment by szpak** [comment]
Doesn't #433 (released in 1.3.2) address that (at least for Maven)?
(https://github.com/hcoles/pitest/pull/460 could be also useful to read)

(for Gradle you can subscribe to https://github.com/szpak/gradle-pitest-plugin/issues/80 to get notification once it is ready)

**Comment by monperrus** [comment]
FYI: Pitmp solves this problem at the Maven level: https://github.com/STAMP-project/pitmp-maven-plugin

**Comment by maxgabut** [comment]
With https://github.com/hcoles/pitest/pull/433 and https://github.com/hcoles/pitest/pull/448 we can now aggregate reports.
